src/shogun/mathematics/linalg/ratapprox/logdet/LogDetEstimator.cpp

```diff
@@ -148,72 +125,81 @@ SGVector<float64_t> CLogDetEstimator::sample(index_t num_estimates)
 		m_operator_log->get_operator()->get_dimension(),
 		m_trace_sampler->get_dimension());
 
-	// for storing the aggregators that submit_jobs return
-	CDynamicObjectArray* aggregators=new CDynamicObjectArray();
+	// for storing the result
+	CDynamicArray<float64_t> aggregators;
 	index_t num_trace_samples=m_trace_sampler->get_num_samples();
-
-	for (index_t i=0; i<num_estimates; ++i)
+	// for omp
+	int32_t num_vectors = num_estimates;
+	auto pb = progress(range(num_vectors));
+	int32_t num_threads;
+	int64_t step;
+#pragma omp parallel shared(num_threads, step)
 	{
-		for (index_t j=0; j<num_trace_samples; ++j)
-		{
-			SG_INFO("Computing log-determinant trace sample %d/%d\n", j,
-					num_trace_samples);
 
-			SG_DEBUG("Creating job for estimate %d, trace sample %d/%d\n", i, j,
+#ifdef HAVE_OPENMP // HAVE OPENMP
+#pragma omp single
+		{
+			num_threads = omp_get_num_threads();
+			step = num_vectors / num_threads;
+			num_threads--;
+		}
+		int32_t thread_num = omp_get_thread_num();
+#else
+		num_threads = 0;
+		step = num_vectors;
+		int32_t thread_num = 0;
+#endif
+		int32_t start = thread_num * step;
+		int32_t end =
+			(thread_num == num_threads) ? num_vectors : (thread_num + 1) * step;
+
+#ifdef WIN32 // HAVE WIN32
+		for (int32_t vec = start; vec < end; vec++)
+#else
+		for (int32_t vec = start; vec < end; vec++)
+#endif
+		{
+			pb.print_progress();
+			for (index_t j = 0; j < num_trace_samples; ++j)
+			{
+				SG_INFO(
+					"Computing log-determinant trace sample %d/%d\n", j,
 					num_trace_samples);
-			// get the trace sampler vector
-			SGVector<float64_t> s=m_trace_sampler->sample(j);
-			// create jobs with the sample vector and store the aggregator
-			CJobResultAggregator* agg=m_operator_log->submit_jobs(s);
-			aggregators->append_element(agg);
-			SG_UNREF(agg);
+				// get the trace sampler vector
+				SGVector<float64_t> s = m_trace_sampler->sample(j);
+				// calculate the result for sample s
+				float64_t agg = m_operator_log->solve(s);
+#pragma omp critical // so that the dynamic array stays concurrent
+				{
+					aggregators.append_element(agg);
```

Comment:
definitely not thread safe. But if you use a vector as mentioned above, this should be ok